### PR TITLE
Join channel

### PR DESCRIPTION
**Pull request by e-stoj**

### CODE REVIEW

File: src/main/java/com/example/chat/groupchatbackend/Conversation.java [hunk 1]
```diff
@@ -71,4 +71,14 @@ public void setConversationType(ConversationType conversationType) {
     public void setUsers(List<User> users) {
         this.users = users;
     }
+
+    public boolean checkIfUserIsInConversation(User user){
```
Comment (michal-marciniec) [inline src/main/java/com/example/chat/groupchatbackend/Conversation.java]: no space between `)` and `{`

File: src/main/java/com/example/chat/groupchatbackend/Conversation.java [hunk 2]
```diff
@@ -71,4 +71,14 @@ public void setConversationType(ConversationType conversationType) {
...
+        for (User currentUser : getUsers()){
```
Comment (michal-marciniec) [inline src/main/java/com/example/chat/groupchatbackend/Conversation.java]: here too

File: src/main/java/com/example/chat/groupchatbackend/Conversation.java [hunk 3]
```diff
@@ -71,4 +71,14 @@ public void setConversationType(ConversationType conversationType) {
...
+            if(currentUser.getId().equals(user.getId())){
```
Comment (michal-marciniec) [inline src/main/java/com/example/chat/groupchatbackend/Conversation.java]: and here too :D

File: src/main/java/com/example/chat/groupchatbackend/Conversation.java (hunk 1 above)
Comment (michal-marciniec) [inline src/main/java/com/example/chat/groupchatbackend/Conversation.java]: let's replace for with stream and `anyMatch` method ;>

File: src/main/java/com/example/chat/groupchatbackend/controllers/ConversationController.java [hunk 4]
```diff
@@ -73,5 +74,30 @@ public ResponseEntity getAllChannelMembers(@PathVariable int id) {
...
+        Conversation conversation = conversationsRepository.findById(id)
+                .orElseThrow(() -> new RuntimeException("conversation doesn't exist"));
+        User currentUser = userContext.getCurrentUser();
+
+        if(conversation.getConversationType().equals(ConversationType.CHANNEL) && currentUser != null){
+            if(!conversation.checkIfUserIsInConversation(currentUser)) {
+                conversation.getUsers().add(currentUser);
+                return new ResponseEntity(HttpStatus.OK);
+            }
+            else {
```
Comment (michal-marciniec) [inline src/main/java/com/example/chat/groupchatbackend/controllers/ConversationController.java]: else in previous line

File: src/main/java/com/example/chat/groupchatbackend/controllers/ConversationController.java [hunk 5]
```diff
@@ -73,5 +74,30 @@ public ResponseEntity getAllChannelMembers(@PathVariable int id) {
...
+                return ResponseEntity
```
Comment (michal-marciniec) [inline src/main/java/com/example/chat/groupchatbackend/controllers/ConversationController.java]: should return OK if a user was newly added or was already in the conversation. ask me for details why if you would like :)

**Review by marcinJer — APPROVED** [review]

**Review by michal-marciniec — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/chat/groupchatbackend/controllers/ConversationController.java [hunk 6]
```diff
@@ -73,5 +74,25 @@ public ResponseEntity getAllChannelMembers(@PathVariable int id) {
...
     }
+
+    @PutMapping("/channels/{id}/join")
+    @Transactional
+    public ResponseEntity joinChannel(@PathVariable int id) {
+        Conversation conversation = conversationsRepository.findById(id)
+                .orElseThrow(() -> new RuntimeException("conversation doesn't exist"));
+        User currentUser = userContext.getCurrentUser();
+
+        if (conversation.getConversationType().equals(ConversationType.CHANNEL) && currentUser != null) {
```
Comment (michal-marciniec) [inline src/main/java/com/example/chat/groupchatbackend/controllers/ConversationController.java]: ` && currentUser != null` is redundant in this context

**Review by michal-marciniec — APPROVED** [review]
First merge #22 into master and then this one ready to go :)

**Pull request merged**